### PR TITLE
feat(api): add host,hostgroup,hostcategory,servicecategory filters to…

### PR DESCRIPTION
**Pull request by adr-mo**
… FindServiceGroups endpoint

## Description

Handle new filters for the `/configuration/services/groups` endpoint
It is now possible to filter with:
- Host Groups
- Service Categories
- Hosts
- Host Categories

Bonus: Get rid of the SqlConcatenator in the modified methods and move back business logic from repository to useCase

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [x] 24.04.x (master)

<h2> How this pull request can be tested ? </h2>
See Jira

#### Community contributors & Centreon team

- [ ] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

## Summary by CodeRabbit

- **New Features**
	- Introduced traits and methods for managing access control to hosts, host categories, host groups, service categories, and service groups based on access group IDs.
	- Added logging for actions related to finding service groups as different user roles.
- **Refactor**
	- Renamed methods and parameters to more accurately reflect their purpose, such as changing `findAllByAccessGroups` to `findAllByAccessGroupIds`.
	- Enhanced SQL query generation for dynamic retrieval of service group data, including support for search parameters, pagination, and sorting.
- **Tests**
	- Updated tests to align with changes in method names, parameters, and the initialization process for service group retrieval.

### CODE REVIEW

**Review by jeremyjaouen — APPROVED** [review]

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 1]
```diff
@@ -64,28 +63,255 @@ public function __construct(DatabaseConnection $db)
...
+    public function findAllByAccessGroupIds(?RequestParametersInterface $requestParameters, array $accessGroupIds): \Traversable&\Countable
     {
-        if ([] === $accessGroups) {
+        if ([] === $accessGroupIds) {
             return new \ArrayIterator([]);
         }
 
-        $accessGroupIds = $this->accessGroupsToIds($accessGroups);
-        if ($this->hasAccessToAllServiceGroups($accessGroupIds)) {
+        [$bindValues, $bindQuery] = $this->createMultipleBindQuery($accessGroupIds, 'access_group_id');
```
Comment (a-launois) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:152]:
```suggestion
        [$bindValues, $bindQuery] = $this->createMultipleBindQuery($accessGroupIds, ':access_group_id');
```

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 2]
```diff
@@ -298,6 +524,257 @@ public function findByIds(int ...$serviceGroupIds): array
...
+    /**
+     * @param int[] $accessGroupIds
+     *
+     * @phpstan-param non-empty-array<int> $accessGroupIds
+     *
+     * @return bool
+     */
+    public function hasAccessToAllHosts(array $accessGroupIds): bool
+    {
+        [$bindValues, $bindQuery] = $this->createMultipleBindQuery($accessGroupIds, 'access_group_id_');
```
Comment (a-launois) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:536]:
```suggestion
        [$bindValues, $bindQuery] = $this->createMultipleBindQuery($accessGroupIds, ':access_group_id_');
```

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 3]
```diff
@@ -298,6 +524,257 @@ public function findByIds(int ...$serviceGroupIds): array
...
+    /**
+     * @param int[] $accessGroupIds
+     *
+     * @phpstan-param non-empty-array<int> $accessGroupIds
+     *
+     * @return bool
+     */
+    public function hasAccessToAllServiceCategories(array $accessGroupIds): bool
+    {
+        [$bindValues, $bindQuery] = $this->createMultipleBindQuery($accessGroupIds, 'access_group_id_');
```
Comment (a-launois) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:574]:
```suggestion
        [$bindValues, $bindQuery] = $this->createMultipleBindQuery($accessGroupIds, ':access_group_id_');
```

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 4]
```diff
@@ -298,6 +524,257 @@ public function findByIds(int ...$serviceGroupIds): array
...
+    /**
+     * @param int[] $accessGroupIds
+     *
+     * @phpstan-param non-empty-array<int> $accessGroupIds
+     *
+     * @return bool
+     */
+    public function hasAccessToAllHostGroups(array $accessGroupIds): bool
+    {
+        [$bindValues, $bindQuery] = $this->createMultipleBindQuery($accessGroupIds, 'access_group_id_');
```
Comment (a-launois) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:608]:
```suggestion
        [$bindValues, $bindQuery] = $this->createMultipleBindQuery($accessGroupIds, ':access_group_id_');
```

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 5]
```diff
@@ -64,28 +63,255 @@ public function __construct(DatabaseConnection $db)
...
+                        INNER JOIN `:db`.acl_resources res
+                            ON res.acl_res_id = arhr.acl_res_id
+                        INNER JOIN `:db`.acl_res_group_relations argr
+                            ON argr.acl_res_id = res.acl_res_id
+                        INNER JOIN `:db`.acl_groups ag
+                            ON ag.acl_group_id = argr.acl_group_id
+                        WHERE ag.acl_group_id IN ({$bindQuery})
+                    )
+                    SQL
+                : '';
```
[the remark this answers is not in the document]
Comment (adr-mo) [reply centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:226]: You are totally right.

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 6]
```diff
@@ -64,28 +63,255 @@ public function __construct(DatabaseConnection $db)
...
+
+            if (\in_array('servicecategory', $matches['object'], true)) {
+                $request .= $this->generateServiceCategoryFilterSubRequest();
+            } elseif (\in_array('hostcategory', $matches['object'], true)) {
+                $request .= $this->generateHostCategoryFilterSubRequest();
+            } elseif (\in_array('hostgroup', $matches['object'], true)) {
+                $request .= $this->generateHostGroupFilterSubRequest();
+            } elseif (\in_array('host', $matches['object'], true)) {
+                $request .= $this->generateHostFilterSubRequest();
+            }
```
[the remark this answers is not in the document]
Comment (adr-mo) [reply centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:83]: You are totally right. Should be usable all at once.

File: centreon/src/Core/ServiceGroup/Application/UseCase/FindServiceGroups/FindServiceGroups.php [hunk 7]
```diff
@@ -94,12 +104,34 @@ private function findServiceGroupAsAdmin(): FindServiceGroupsResponse
...
+            $this->readAccessGroupRepository->findByContact($this->contact)
         );
 
+        if ($accessGroupIds === []) {
+            return $this->createResponse($serviceGroups);
+        }
+
+        if ($this->readServiceGroupRepository->hasAccessToAllServiceGroups($accessGroupIds)) {
+            $this->debug(
+                'ACL configuration for user gives access to all hostgroups',
```
Comment (callapa) [inline centreon/src/Core/ServiceGroup/Application/UseCase/FindServiceGroups/FindServiceGroups.php:120]:
```suggestion
                'ACL configuration for user gives access to all servicegroups',
```

File: centreon/src/Core/ServiceGroup/Application/UseCase/FindServiceGroups/FindServiceGroups.php [hunk 8]
```diff
@@ -94,12 +104,34 @@ private function findServiceGroupAsAdmin(): FindServiceGroupsResponse
...
+                ['user' => $this->contact->getName()]
+            );
+            $serviceGroups = $this->readServiceGroupRepository->findAll($this->requestParameters);
+        } else {
+            $this->debug(
+                'Using users ACL configured on hostgroups',
```
Comment (callapa) [inline centreon/src/Core/ServiceGroup/Application/UseCase/FindServiceGroups/FindServiceGroups.php:126]:
```suggestion
                'Using users ACL configured on servicegroups',
```

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 9]
```diff
@@ -298,6 +487,318 @@ public function findByIds(int ...$serviceGroupIds): array
         return $serviceGroups;
     }
 
+    /**
+     * @param int[] $accessGroupIds
+     *
+     * @phpstan-param non-empty-array<int> $accessGroupIds
```
Comment (callapa) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:493]: This rule is not managed by this method.
Remove this phpstan rule and simply check for the presence of an empty array in the method.

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 10]
```diff
@@ -298,6 +487,318 @@ public function findByIds(int ...$serviceGroupIds): array
...
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * @param int[] $accessGroupIds
+     *
+     * @phpstan-param non-empty-array<int> $accessGroupIds
```
Comment (callapa) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:531]: Same comment as above.

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 11]
```diff
@@ -298,6 +487,318 @@ public function findByIds(int ...$serviceGroupIds): array
...
+
+        $statement->execute();
+
+        return (bool) $statement->fetchColumn();
+    }
+
+    /**
+     * @param int[] $accessGroupIds
+     *
+     * @phpstan-param non-empty-array<int> $accessGroupIds
```
Comment (callapa) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:565]: Same comment as above.

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 12]
```diff
@@ -298,6 +487,318 @@ public function findByIds(int ...$serviceGroupIds): array
...
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * @param int[] $accessGroupIds
+     *
+     * @phpstan-param non-empty-array<int> $accessGroupIds
```
Comment (callapa) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:603]: Same comment as above.

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 13]
```diff
@@ -298,6 +487,318 @@ public function findByIds(int ...$serviceGroupIds): array
...
+
+        $statement->execute();
+
+        return (bool) $statement->fetchColumn();
+    }
+
+    /**
+     * @param int[] $accessGroupIds
+     *
+     * @phpstan-param non-empty-array<int> $accessGroupIds
```
Comment (callapa) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:637]: Same comment as above.

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 14]
```diff
@@ -298,6 +487,318 @@ public function findByIds(int ...$serviceGroupIds): array
...
+     *
+     * @return bool
+     */
+    public function hasAccessToAllHostCategories(array $accessGroupIds): bool
```
Comment (callapa) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:607]: This method must be in the HostCategory repository or in a HostCategory repository trait..

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 15]
```diff
@@ -298,6 +487,318 @@ public function findByIds(int ...$serviceGroupIds): array
...
+    }
+
+    /**
+     * @param int[] $accessGroupIds
+     *
+     * @phpstan-param non-empty-array<int> $accessGroupIds
+     *
+     * @return bool
+     */
+    public function hasAccessToAllHostGroups(array $accessGroupIds): bool
```
Comment (callapa) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:569]: This method must be in the HostGroup repository or in a hostgroup repository trait.

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 16]
```diff
@@ -298,6 +487,318 @@ public function findByIds(int ...$serviceGroupIds): array
...
+    }
+
+    /**
+     * @param int[] $accessGroupIds
+     *
+     * @phpstan-param non-empty-array<int> $accessGroupIds
+     *
+     * @return bool
+     */
+    public function hasAccessToAllServiceCategories(array $accessGroupIds): bool
```
Comment (callapa) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:535]: This method must be in the ServiceCategory repository or in a ServiceCategory repository trait.

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 17]
```diff
@@ -298,6 +487,318 @@ public function findByIds(int ...$serviceGroupIds): array
...
     }
 
+    /**
+     * @param int[] $accessGroupIds
+     *
+     * @phpstan-param non-empty-array<int> $accessGroupIds
+     *
+     * @return bool
+     */
+    public function hasAccessToAllHosts(array $accessGroupIds): bool
```
Comment (callapa) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:497]: This method must be in the Host repository or in a Host repository trait

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 18]
```diff
@@ -298,6 +487,318 @@ public function findByIds(int ...$serviceGroupIds): array
...
+
+        return false;
+    }
+
+    /**
+     * @param int[] $accessGroupIds
+     *
+     * @return string
+     */
+    private function generateHostAclSubRequest(array $accessGroupIds = []): string
```
Comment (callapa) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:677]: This method must be in the Host repository or in a host repository trait.

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 19]
```diff
@@ -298,6 +487,318 @@ public function findByIds(int ...$serviceGroupIds): array
...
+
+        return $subRequest;
+    }
+
+    /**
+     * @param int[] $accessGroupIds
+     *
+     * @return string
+     */
+    private function generateServiceCategoryAclSubRequest(array $accessGroupIds = []): string
```
Comment (callapa) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:710]: This method must be in the Host repository or in a ServiceCategory repository trait.

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 20]
```diff
@@ -298,6 +487,318 @@ public function findByIds(int ...$serviceGroupIds): array
...
+
+        return $request;
+    }
+
+    /**
+     * @param int[] $accessGroupIds
+     *
+     * @return string
+     */
+    private function generateHostCategoryAclSubRequest(array $accessGroupIds = []): string
```
Comment (callapa) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:743]: This method must be in the Host repository or in a Hostcategory repository trait.

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 21]
```diff
@@ -298,6 +487,318 @@ public function findByIds(int ...$serviceGroupIds): array
...
+
+        return $request;
+    }
+
+    /**
+     * @param int[] $accessGroupIds
+     *
+     * @return string
+     */
+    private function generateHostGroupAclSubRequest(array $accessGroupIds = []): string
```
Comment (callapa) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:776]: This method must be in the Host repository or in a HostGroup repository trait.

**Review by callapa — CHANGES_REQUESTED** [review]

**Review by callapa — DISMISSED** [review]

**Review by callapa — APPROVED** [review]

**Review by jeremyjaouen — APPROVED** [review]

**Pull request merged**